File: python/smoke.py
"""Smoke test for the Python `amber` wrapper (Plans.md 10.4).

Validates that `import amber` works and exposes the capture-once surface. Run
against a built/generated binding (the maturin wheel, `maturin develop`, or a
`uniffi-bindgen`-generated `amber_core.py` + cdylib on `sys.path`):

    AMBER_CHROMIUM_PATH="$(command -v chromium || true)" python python/smoke.py

A data: URL keeps it self-contained; PDF/screenshot exercise a real browser, so
set AMBER_CHROMIUM_PATH (or let the pinned Chrome for Testing download once).
"""

import sys

import amber

URL = "data:text/html,<html><body><h1>Smoke</h1><p>hello</p></body></html>"


def main() -> int:
    md = amber.capture_markdown(URL)
    assert "Smoke" in md, f"markdown missing content: {md!r}"

    # Capture once, emit many — one render serves every format.
    snap = amber.snapshot(URL, [amber.OutputFormat.MARKDOWN, amber.OutputFormat.PDF])
    assert "Smoke" in snap.markdown(), "snapshot markdown missing content"
    assert snap.render(amber.OutputFormat.PDF)[:4] == b"%PDF", "snapshot not a PDF"

    try:
        amber.snapshot("not a url", [amber.OutputFormat.MARKDOWN])
    except amber.CaptureError:
        pass  # expected
    else:
        raise AssertionError("expected CaptureError for a bad URL")

    print("python amber smoke OK")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: bindings/ruby/test/smoke.rb
# frozen_string_literal: true

# Smoke test for the Ruby binding. Run `bindings/ruby/generate.sh` first, then:
#   ruby bindings/ruby/test/smoke.rb
$LOAD_PATH.unshift(File.expand_path("../lib", __dir__))
require "amber_html"

url = "data:text/html,<html><body><h1>Smoke</h1><p>hello</p></body></html>"

md = AmberHtml.capture_markdown(url)
raise "markdown missing content" unless md.include?("Smoke")

pdf = AmberHtml.capture(url, AmberHtml::OutputFormat::PDF)
raise "not a PDF" unless pdf[0, 4] == "%PDF"

png = AmberHtml.capture(url, AmberHtml::OutputFormat::SCREENSHOT)
raise "not a PNG" unless png[1, 3] == "PNG"

begin
  AmberHtml.capture_markdown("not a url")
  raise "expected an error for a bad URL"
rescue AmberHtml::CaptureError::Failed
  # expected — the facade surfaces capture failures as CaptureError::Failed
end

# Capture-once Snapshot object (Plans.md 10.1): one capture serves many formats.
require "tmpdir"
snap = AmberHtml.snapshot(url, [AmberHtml::OutputFormat::MARKDOWN, AmberHtml::OutputFormat::PDF])
raise "snapshot markdown missing content" unless snap.markdown.include?("Smoke")
raise "snapshot not a PDF" unless snap.render(AmberHtml::OutputFormat::PDF)[0, 4] == "%PDF"
saved = snap.save(AmberHtml::OutputFormat::READABLE, Dir.tmpdir, "amber_ruby_snap")
raise "snapshot save failed" unless File.exist?(saved)

begin
  AmberHtml.snapshot("not a url", [AmberHtml::OutputFormat::MARKDOWN])
  raise "expected an error for a bad URL (snapshot)"
rescue AmberHtml::CaptureError::Failed
  # expected
end

puts "ruby smoke OK (markdown #{md.bytesize}B, pdf #{pdf.bytesize}B, png #{png.bytesize}B, " \
     "snapshot #{snap.markdown.bytesize}B md)"
